### PR TITLE
[test] Implement rom_e2e_ret_ram_init_test

### LINKED ISSUE #14494: [rom-e2e] rom_e2e_ret_ram_init
**Testpoint name:** [rom_e2e_ret_ram_init](https://cs.opensource.google/opentitan/opentitan/+/master:sw/device/silicon_creator/rom/data/rom_e2e_testplan.hjson?q=rom_e2e_ret_ram_init)

**Contact person:** @alphan

**Description:** Verify that ROM initializes the retention SRAM when needed.

`CREATOR_SW_CFG_RET_RAM_RESET_MASK` should be `1 << 3` so that ROM initializes the
retention SRAM after a SW reset request.

- Verify that `reset_reasons` reports POR.
- Write a known pattern into all sections of retention SRAM.
- Request a SW reset.
- Verify that `reset_reasons` reports a SW request.
- Verify that all previously written sections are different.

### PR DESCRIPTION
**Pull request by milesdai**
Closes #14494 
Signed-off-by: Miles Dai <milesdai@google.com>

### CODE REVIEW

File: sw/device/silicon_creator/rom/e2e/BUILD [hunk 1]
```diff
@@ -2449,3 +2449,59 @@ test_suite(
     tags = ["manual"],
     tests = ["sigverify_usage_constraint_{}".format(t["name"]) for t in test_cases],
 )
+
+################################################################
+# Big comment here to prevent Git merge from
+# interleaving during a merge
+# TODO: remove this guard comment
```
Comment (milesdai) [inline sw/device/silicon_creator/rom/e2e/BUILD:2456]: TODO: remove this comment

File: sw/device/silicon_creator/rom/e2e/rom_e2e_ret_ram_init_test.c [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+#include "sw/device/silicon_creator/lib/drivers/rstmgr.h"
+#include "sw/device/silicon_creator/lib/test_main.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+OTTF_DEFINE_TEST_CONFIG();
+
+// Variables of type `retention_sram_t` are static to reduce stack usage.
+static retention_sram_t ret;
+static uint64_t raw[sizeof(retention_sram_t) / sizeof(uint64_t)];
```
Comment (alphan) [inline sw/device/silicon_creator/rom/e2e/rom_e2e_ret_ram_init_test.c:19]: nit: can you either (1) add an assert that `sizeof(retention_sram_t)` is a multiple of `sizeof(uint64_t)`, or (2) just add `+1` to the number of elements to make sure that we don't do OOB access.

File: sw/device/silicon_creator/rom/e2e/rom_e2e_ret_ram_init_test.c [hunk 3]
```diff
@@ -0,0 +1,62 @@
...
+  if (bitfield_bit32_read(reset_reasons, kRstmgrReasonPowerOn)) {
+    // This branch runs after the POR after initializing the testing environment
+    LOG_INFO("Writing known pattern");
+    memset(ret, 0xab, sizeof(retention_sram_t));
+
+    LOG_INFO("Requesting SW reset");
+    dif_rstmgr_t rstmgr;
+    CHECK_DIF_OK(dif_rstmgr_init(
+        mmio_region_from_addr(TOP_EARLGREY_RSTMGR_AON_BASE_ADDR), &rstmgr));
+    CHECK_DIF_OK(dif_rstmgr_software_device_reset(&rstmgr));
```
Comment (alphan) [inline sw/device/silicon_creator/rom/e2e/rom_e2e_ret_ram_init_test.c:35]: can you use the `rstmgr` driver?

File: sw/device/silicon_creator/rom/e2e/rom_e2e_ret_ram_init_test.c [hunk 4]
```diff
@@ -0,0 +1,62 @@
...
+  } else if (bitfield_bit32_read(reset_reasons, kRstmgrReasonSoftwareRequest)) {
+    // This branch runs after the SW-requested reset
+    LOG_INFO("Ensuring all sections have changed");
+    memcpy(raw, ret, sizeof(retention_sram_t));
+
+    uint32_t matches = 0;
+    for (size_t i = 0; i < ARRAYSIZE(raw); ++i) {
+      if (raw[i] == 0xabababababababab) {
```
Comment (alphan) [inline sw/device/silicon_creator/rom/e2e/rom_e2e_ret_ram_init_test.c:43]: nit: can you define this forbidden value like the following?

```
uint64_t pattern64;
memset(&pattern64, kPattern, sizeof(pattern64));
```

File: sw/device/silicon_creator/rom/e2e/rom_e2e_ret_ram_init_test.c [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+#include "sw/device/lib/testing/test_framework/check.h"
+#include "sw/device/lib/testing/test_framework/ottf_main.h"
+#include "sw/device/silicon_creator/lib/drivers/retention_sram.h"
+#include "sw/device/silicon_creator/lib/drivers/rstmgr.h"
+#include "sw/device/silicon_creator/lib/test_main.h"
+
+#include "hw/top_earlgrey/sw/autogen/top_earlgrey.h"
+
+OTTF_DEFINE_TEST_CONFIG();
+
```
Comment (alphan) [inline sw/device/silicon_creator/rom/e2e/rom_e2e_ret_ram_init_test.c:16]:
```suggestion
enum {
    kPattern = 0xab,
};
```

File: sw/device/silicon_creator/rom/e2e/rom_e2e_ret_ram_init_test.c [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+rom_error_t retention_ram_init_test(void) {
+  retention_sram_t *ret = retention_sram_get();
+  uint32_t reset_reasons = ret->reset_reasons;
+
+  // Verify that reset_reasons reports POR.
+  if (bitfield_bit32_read(reset_reasons, kRstmgrReasonPowerOn)) {
+    // This branch runs after the POR after initializing the testing environment
+    LOG_INFO("Writing known pattern");
+    memset(ret, 0xab, sizeof(retention_sram_t));
```
Comment (alphan) [inline sw/device/silicon_creator/rom/e2e/rom_e2e_ret_ram_init_test.c:29]:
```suggestion
    memset(ret, kPattern, sizeof(retention_sram_t));
```

**Review by alphan — APPROVED** [review]
Thanks @milesdai !

File: sw/device/silicon_creator/rom/e2e/rom_e2e_ret_ram_init_test.c (hunk 3 above)
Comment (milesdai) [reply sw/device/silicon_creator/rom/e2e/rom_e2e_ret_ram_init_test.c]: Yup -- that was much easier!

**Review by alphan — APPROVED** [review]

**Comment by alphan** [comment]
Rebased.

**Pull request merged**